test/model/compose_test.dart

```diff
@@ -0,0 +1,391 @@
+```javascript
+// 456
+```
+````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('whitespace around info string', () {
+      final content = chomp('''
+``` javascript 
+// hello world
+```
+''');
+      final expected = chomp('''
+````
+``` javascript 
+// hello world
+```
+````
+''');
+      final expectedWithInfoString = chomp('''
+````info
+``` javascript 
+// hello world
+```
+````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('four-backtick block', () {
+      final content = chomp('''
+````
+hello world
+````
+''');
+      final expected = chomp('''
+`````
+````
+hello world
+````
+`````
+''');
+      final expectedWithInfoString = chomp('''
+`````info
+````
+hello world
+````
+`````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('five-backtick block', () {
+      final content = chomp('''
+`````
+hello world
+`````
+''');
+      final expected = chomp('''
+``````
+`````
+hello world
+`````
+``````
+''');
+      final expectedWithInfoString = chomp('''
+``````info
+`````
+hello world
+`````
+``````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('three-, four-, and five-backtick blocks', () {
+      final content = chomp('''
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+''');
+      final expected = chomp('''
+``````
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+``````
+''');
+      final expectedWithInfoString = chomp('''
+``````info
+```
+hello world
+```
+
+````
+hello world
+````
+
+`````
+hello world
+`````
+``````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('dangling opening fence', () {
+      final content = chomp('''
+````javascript
+// hello world
+''');
+      final expected = chomp('''
+`````
+````javascript
+// hello world
+`````
+''');
+      final expectedWithInfoString = chomp('''
+`````info
+````javascript
+// hello world
+`````
+''');
+      check(wrapWithBacktickFence(content: content)).equals(expected);
+      check(wrapWithBacktickFence(content: content, infoString: 'info')).equals(expectedWithInfoString);
+    });
+
+    test('no panic on badly positioned fences', () {
+      final content = chomp('''
+````javascript
+// hello world
+```
+asdf
+````
+```python
+''');
```

Comment:
Hmm, I wouldn't even call these "bad" in any way.

There's a "javascript" code block containing:
````
// hello world
```
asdf
````
followed by a "python" code block that's empty.

That isn't well-formed JS inside the "javascript" block, but the Markdown here is perfectly fine — certainly according to CommonMark and to the Zulip server.